elyra/pipeline/airflow/component_parser_airflow.py

```diff
@@ -46,149 +39,316 @@ class AirflowComponentParser(ComponentParser):
-            match = class_regex.search(line)
-            if match:
-                class_name = match.group(1)
-                class_to_content[class_name] = {"lines": [], "args": []}
-            class_to_content[class_name]['lines'].append(line)
-
-        class_to_content.pop("no_class")
-
-        init_regex = re.compile(r"def __init__\(([\s\d\w,=\-\'\"\[\]\{\}\*\s\#.\\\/:?]*)\):")
-        for class_name, content in class_to_content.items():
-            # Concatenate class body and search for __init__ function
-            class_content = '\n'.join(content['lines'])
-            # remove type hint via TypeHintRemover with ast parse
-            parsed_source = ast.parse(class_content)
-            transformed = TypeHintRemover().visit(parsed_source)
-            class_content = astunparse.unparse(transformed)
-
-            for match in init_regex.finditer(class_content):
-                # Get list of parameter:default-value pairs
-                class_to_content[class_name]['args'] = [x.strip() for x in match.group(1).split(',') if x.strip() != '']
-
-        return class_to_content
-
-    def _parse_properties(self, content: Dict[str, List]) -> List[ComponentParameter]:
-        properties: List[ComponentParameter] = list()
+    def _filter_operator_classes(self,
+                                 class_def_nodes: List[ast.ClassDef],
+                                 import_from_nodes: List[ast.ImportFrom]) -> List[ast.ClassDef]:
+        """
+        Analyze each ast.ClassDef object to determine whether it directly or indirectly
+        extends the BaseOperator.
+
+        :param class_def_nodes: a list of all ast.ClassDef objects in a file
+        :param import_from_nodes: a list of all ast.ImportFrom objects in a file (only used to satisfy
+                                  the temporary workaround called out below)
+
+        :returns: a filtered list of ast.ClassDef objects that can be considered 'Operators'
+        """
+        operator_classes = []
+        classes_to_analyze = []
+
+        # Get class names for package imports that match one of the following patterns,
+        # indicating that this class does match a known Operator class as defined in
+        # a provider package or core Airflow package
+        regex_patterns = [
+            re.compile(r'airflow\.providers\.[a-z_]+\.operators'),  # airflow.providers.*.operators. (provider package)
+            re.compile(r'airflow\.operators\.')  # airflow.operators.* (core Airflow package)
+        ]
+        operator_bases = ['BaseOperator']
+        for module in import_from_nodes:
+            if any(regex.match(module.module) for regex in regex_patterns):
+                operator_bases.extend([name.name for name in module.names])
+
+        # Determine whether each class directly extends the BaseOperator or whether it
+        # must be further analyzed for indirect extension
+        for node in class_def_nodes:
+            if len(node.bases) == 0:
```

Comment:
Can `node.bases` be `None`?  If so, it might be safer to ask `if not node.bases:`

Reply:
Good point. I haven't seen a scenario where it's `None` yet, but better safe than sorry!